### DISCUSSION TITLE
Should I always give output values ​​in the evaluate function?

### DISCUSSION THREAD

**Discussion by ubeydtalha** [anyoptimization/pymoo#213] [Q&A] [answered]
First of all, I would like to thank everyone who contributed to this library. ^^

I'm developing a circuit optimization tool, simulating the input values ​​in the circuit and passing the output values ​​to pymoo.

But a numerical value does not always come from the circuit simulation, the circuit may give an error according to the individuals produced.

In this case, the value is "failed" and I cannot output this value.(out["F"])

**In this case, is it sufficient to just terminate the function with exiting with "return"?**

Nor can I define a constraint function and have it fail manually, as I have to use the MOEA/D algorithm.

**Comment by blankjul** [comment]
This looks like a hidden constraint to me with failed runs. The easiest fix is to wrap your function with `try...except` and set the objective value(s) to `np.inf` whenever it fails.
Let me know if this works for you.

**Reply by ubeydtalha** [reply]
I ended the function with return without giving any value to the purposes.but I don't know how true?
I will try what you said ^^

````python


    def _evaluate(self, x, out, *args, **kwargs):

        meAbsPath = os.path.dirname(os.path.realpath(__file__))
        h = HSpicePy(file_name="amp",design_file_name="designparam",path=meAbsPath,timeout="")
        h.change_parameters_to_cir(LM1 = x[0], LM2 = x[1], LM3 = x[2],
                        WM1 = x[3], WM2 = x[4], WM3 = x[5],
                        Rb = x[6],Vb = x[7])
        asyncio.run(h.run_async())
        # h.run()
        f1_ = h.result["bw"]
        f2_ = h.result["gain"]
        
        if "failed" in [f1_,f2_]:
            return
            out["F"] = [0, 0]
            
        else:
            f1 = float(h.result["bw"])
            f2 = -float(h.result["gain"])

            out["F"] = [f1, f2]
```

**Comment by blankjul** [comment, marked as answer]
just set `out["F"] = [np.inf, np.inf]` when it fails.

**Reply by ubeydtalha** [reply]
thank you ^^

**Reply by ubeydtalha** [reply]
Why can't we add out["G"] to the MOEA/D algorithm?

**Reply by blankjul** [reply]
the MOEA/D  algorithm does not have any constraint handling (not part of the original paper)

**Reply by ubeydtalha** [reply]
Got it, thank you very much for your help.

**Reply by ubeydtalha** [reply]
> just set `out["F"] = [np.inf, np.inf]` when it fails.

hello, when you do as you say, x values ​​start to be created in true-false, not float.

To prevent this, when I write return without giving out["F"] to the _eveluate function, I get the this [error](https://github.com/anyoptimization/pymoo/discussions/217)

![image](https://user-images.githubusercontent.com/62707118/142252374-5212a51d-e6b7-447b-9a02-c6fb760081da.png)

**Reply by blankjul** [reply]
I would not no of any reason why setting "F", should impact the "X" of any iteration in the future. Have you figured it out?
